rfcs/0074-tvmscript-unified-printer.md

```diff
@@ -0,0 +1,487 @@
+
+Error: ...
+```
+
+However, the underlying mechanism supports only S-TIR and error reporting on
+`tir.ForNode` and `tir.BlockNode`, and is less extensible for generic cases.
+
+To generalize this UX across the TVM stack, during the first stage in
+translation, the following steps is additionally executed:
+
+- Each object in the Doc DSL is optionally attached to a node in TVM IR
+- After the 1st stage is finished, collect all IR nodes that gets attached to
+  Doc into a map, whose key is IR node and value is a list of Doc nodes.
+- For each IR node that has diagnostic message, trace back through its parent
+  until it reaches to an IR node in the map collected in previous step. Then it
+  can produce a map from Doc node to diagnostic message.
+- In the 2nd stage, diagnostic message will be printed as doc is being printed
+  into target language
+
+# Reference-level explanation
+
+## Doc DSL Spec
+
+The design of the Doc DSL is to have a unified representation of TVMScript in
+different languages. The overall structure is simplied from Python ast, and
+their meaning is straightforward.
+
+```py
+Doc(Optional<ObjectRef> source) # Base class for doc
+
+# Expression
+ExprDoc() # Base class for expression
+LiteralDoc(Union[IntImm, FloatImm, String, nullptr_t] value) 
+IdDoc(String name)
+AttrAccessDoc(ExprDoc value, String attr)
+IndexDoc(ExprDoc value, Array<Union<ExprDoc, SliceDoc>> indices) 
+CallDoc(ExprDoc callee, Array<ExprDoc> args, Array<String> kwargs_keys, Array<ExprDoc> kwargs_values)
+OperationDoc(OperationKind kind, Array<ExprDoc> operands)
+LambdaDoc(Array<IdDoc> args, ExprDoc body)
+TupleDoc(Array<ExprDoc> elements)
+ListDoc(Array<ExprDoc> elements)
+DictDoc(Array<ExprDoc> keys, Array<ExprDoc> values)
+
+# Statements
+StmtDoc(Array<String> comments) # Base class
+AssignDoc(ExprDoc lhs, Optional<ExprDoc> rhs, Optional<ExprDoc> annotation)
+IfDoc(ExprDoc predicate, Array<StmtDoc> then_branch, Array<StmtDoc> else_branch)
+WhileDoc(ExprDoc predicate, Array<StmtDoc> body)
+ForDoc(ExprDoc lhs, ExprDoc rhs, Array<StmtDoc> body)
+ScopeDoc(Optional<ExprDoc> lhs, ExprDoc rhs, Array<StmtDoc> body)
+ExprStmtDoc(ExprDoc expr)
+
+# Special Docs
+SliceDoc(Optional<ExprDoc> start, Optional<ExprDoc> stop)
+FunctionDoc(IdDoc name, Array<AssignDoc> args, Array<ExprDoc> decorators, ExprDoc return_type, Array<StmtDoc> body))
+ClassDoc(IdDoc name, Array<ExprDoc> decorators, Array<AssignDoc> aliases, Array<FunctionDoc> functions)
+```
+
+## IRDocsifier Spec
+
+IRDocsifier is responsible for transforming IR node tree into Doc tree. Its API
+looks like
+
+```cpp
+class IRDocsifierNode : public Object {
+ public:
+  // ir_prefix maintains a map from dispatch_token to ir prefix
+  // so that the print function can construct an expression with
+  // the current ir prefix, like `T.xxx` in TIR and `R.xxx` in Relax 
+  Map<String, String> ir_prefix;
+  // SymbolTable maintains the mapping from IR node to Doc
+  SymbolTable sym;
```

Comment:
maybe find a better name than symbol table?

Reply:
Updated to `TranslationTable`.. Not sure if it's better